File: patterns/_tests/BehavioralPatterns/Interpreter/csharp/unit.test.cs
// ============================================================
//  Unit Tests — Interpreter: Arithmetic Expression Evaluator (C#)
//
//  Run: dotnet test
// ============================================================

using Xunit;
using System;
using System.Collections.Generic;

// ── Context ───────────────────────────────────────────────────
public class ContextTests {
    [Fact]
    public void StoresAndRetrievesVariable() {
        var ctx = new Context().Set("x", 42);
        Assert.Equal(42, ctx.Get("x"));
    }

    [Fact]
    public void ThrowsOnMissingVariable() {
        var ctx = new Context();
        Assert.Throws<KeyNotFoundException>(() => ctx.Get("missing"));
    }

    [Fact]
    public void OverwritesExistingVariable() {
        var ctx = new Context().Set("x", 1).Set("x", 99);
        Assert.Equal(99, ctx.Get("x"));
    }

    [Fact]
    public void SetReturnsSelfForChaining() {
        var ctx = new Context();
        Assert.Same(ctx, ctx.Set("a", 1));
    }
}


// ── NumberExpression ──────────────────────────────────────────
public class NumberExpressionTests {
    [Fact]
    public void ReturnsLiteralValue() {
        var expr = new NumberExpression(7);
        Assert.Equal(7, expr.Interpret(new Context()));
    }

    [Fact]
    public void IgnoresContext() {
        var ctx  = new Context().Set("x", 999);
        var expr = new NumberExpression(3);
        Assert.Equal(3, expr.Interpret(ctx));
    }

    [Fact]
    public void DisplayRendersInteger() {
        Assert.Equal("5", new NumberExpression(5).Display());
    }
}


// ── VariableExpression ────────────────────────────────────────
public class VariableExpressionTests {
    [Fact]
    public void ResolvesFromContext() {
        var ctx  = new Context().Set("a", 10);
        var expr = new VariableExpression("a");
        Assert.Equal(10, expr.Interpret(ctx));
    }

    [Fact]
    public void ThrowsWhenNotInContext() {
        var expr = new VariableExpression("z");
        Assert.Throws<KeyNotFoundException>(() => expr.Interpret(new Context()));
    }

    [Fact]
    public void DisplayReturnsName() {
        Assert.Equal("a", new VariableExpression("a").Display());
    }
}


// ── AddExpression ─────────────────────────────────────────────
public class AddExpressionTests {
    [Fact]
    public void AddsTwoNumbers() {
        var expr = new AddExpression(new NumberExpression(3), new NumberExpression(4));
        Assert.Equal(7, expr.Interpret(new Context()));
    }

    [Fact]
    public void AddsTwoVariables() {
        var ctx  = new Context().Set("a", 10).Set("b", 5);
        var expr = new AddExpression(new VariableExpression("a"), new VariableExpression("b"));
        Assert.Equal(15, expr.Interpret(ctx));
    }

    [Fact]
    public void DisplayContainsPlusAndParens() {
        var expr = new AddExpression(new NumberExpression(1), new NumberExpression(2));
        Assert.Contains("+", expr.Display());
        Assert.StartsWith("(", expr.Display());
    }
}


// ── SubtractExpression ────────────────────────────────────────
public class SubtractExpressionTests {
    [Fact]
    public void SubtractsRightFromLeft() {
        var expr = new SubtractExpression(new NumberExpression(10), new NumberExpression(3));
        Assert.Equal(7, expr.Interpret(new Context()));
    }

    [Fact]
    public void ResultCanBeNegative() {
        var expr = new SubtractExpression(new NumberExpression(2), new NumberExpression(5));
        Assert.Equal(-3, expr.Interpret(new Context()));
    }
}


// ── MultiplyExpression ────────────────────────────────────────
public class MultiplyExpressionTests {
    [Fact]
    public void MultipliesTwoNumbers() {
        var expr = new MultiplyExpression(new NumberExpression(4), new NumberExpression(5));
        Assert.Equal(20, expr.Interpret(new Context()));
    }

    [Fact]
    public void MultipliesVariableByNumber() {
        var ctx  = new Context().Set("a", 6);
        var expr = new MultiplyExpression(new VariableExpression("a"), new NumberExpression(7));
        Assert.Equal(42, expr.Interpret(ctx));
    }
}


// ── DivideExpression ──────────────────────────────────────────
public class DivideExpressionTests {
    [Fact]
    public void DividesLeftByRight() {
        var expr = new DivideExpression(new NumberExpression(10), new NumberExpression(2));
        Assert.Equal(5, expr.Interpret(new Context()));
    }

    [Fact]
    public void ThrowsOnDivisionByZero() {
        var expr = new DivideExpression(new NumberExpression(5), new NumberExpression(0));
        Assert.Throws<DivideByZeroException>(() => expr.Interpret(new Context()));
    }

    [Fact]
    public void ProducesFractionalResult() {
        var expr = new DivideExpression(new NumberExpression(10), new NumberExpression(4));
        Assert.Equal(2.5, expr.Interpret(new Context()));
    }
}


// ── Composite trees ───────────────────────────────────────────
public class CompositTreeTests {
    [Fact]
    public void DiffOfSquares() {
        IExpression a   = new VariableExpression("a");
        IExpression b   = new VariableExpression("b");
        var ctx = new Context().Set("a", 10).Set("b", 5);
        var expr = new MultiplyExpression(new AddExpression(a, b), new SubtractExpression(a, b));
        Assert.Equal(75, expr.Interpret(ctx));
    }

    [Fact]
    public void SameTreeDifferentContexts() {
        IExpression a    = new VariableExpression("a");
        IExpression b    = new VariableExpression("b");
        var expr = new MultiplyExpression(new AddExpression(a, b), new SubtractExpression(a, b));
        Assert.Equal(75, expr.Interpret(new Context().Set("a", 10).Set("b", 5)));
        Assert.Equal(32, expr.Interpret(new Context().Set("a",  6).Set("b", 2)));
    }

    [Fact]
    public void LiteralThreeLevelTree() {
        var expr = new MultiplyExpression(
            new AddExpression(new NumberExpression(3), new NumberExpression(4)),
            new NumberExpression(2));
        Assert.Equal(14, expr.Interpret(new Context()));
    }
}

File: patterns/_tests/BehavioralPatterns/Interpreter/java/unit.test.java
// ============================================================
//  Unit Tests — Interpreter: Arithmetic Expression Evaluator (Java)
//
//  Run: javac java.java unit.test.java && java org.junit.runner.JUnitCore InterpreterUnitTests
// ============================================================

import org.junit.Test;
import org.junit.Before;
import static org.junit.Assert.*;
import java.util.NoSuchElementException;

public class InterpreterUnitTests {

    // ── Context ───────────────────────────────────────────────
    @Test
    public void context_storesAndRetrievesVariable() {
        Context ctx = new Context().set("x", 42);
        assertEquals(42.0, ctx.get("x"), 0.001);
    }

    @Test(expected = NoSuchElementException.class)
    public void context_throwsOnMissingVariable() {
        new Context().get("missing");
    }

    @Test
    public void context_overwritesExistingVariable() {
        Context ctx = new Context().set("x", 1).set("x", 99);
        assertEquals(99.0, ctx.get("x"), 0.001);
    }

    @Test
    public void context_setReturnsSelf() {
        Context ctx = new Context();
        assertSame(ctx, ctx.set("a", 1));
    }


    // ── NumberExpression ──────────────────────────────────────
    @Test
    public void number_returnsLiteralValue() {
        assertEquals(7.0, new NumberExpression(7).interpret(new Context()), 0.001);
    }

    @Test
    public void number_ignoresContext() {
        Context ctx = new Context().set("x", 999);
        assertEquals(3.0, new NumberExpression(3).interpret(ctx), 0.001);
    }

    @Test
    public void number_displayRendersInteger() {
        assertEquals("5", new NumberExpression(5).display());
    }


    // ── VariableExpression ────────────────────────────────────
    @Test
    public void variable_resolvesFromContext() {
        Context ctx = new Context().set("a", 10);
        assertEquals(10.0, new VariableExpression("a").interpret(ctx), 0.001);
    }

    @Test(expected = NoSuchElementException.class)
    public void variable_throwsWhenNotInContext() {
        new VariableExpression("z").interpret(new Context());
    }

    @Test
    public void variable_displayReturnsName() {
        assertEquals("a", new VariableExpression("a").display());
    }


    // ── AddExpression ─────────────────────────────────────────
    @Test
    public void add_addsTwoNumbers() {
        Expression expr = new AddExpression(new NumberExpression(3), new NumberExpression(4));
        assertEquals(7.0, expr.interpret(new Context()), 0.001);
    }

    @Test
    public void add_addsTwoVariables() {
        Context ctx = new Context().set("a", 10).set("b", 5);
        Expression expr = new AddExpression(new VariableExpression("a"), new VariableExpression("b"));
        assertEquals(15.0, expr.interpret(ctx), 0.001);
    }

    @Test
    public void add_displayContainsPlusAndParens() {
        Expression expr = new AddExpression(new NumberExpression(1), new NumberExpression(2));
        assertTrue(expr.display().contains("+"));
        assertTrue(expr.display().startsWith("("));
    }


    // ── SubtractExpression ────────────────────────────────────
    @Test
    public void subtract_subtractsRightFromLeft() {
        Expression expr = new SubtractExpression(new NumberExpression(10), new NumberExpression(3));
        assertEquals(7.0, expr.interpret(new Context()), 0.001);
    }

    @Test
    public void subtract_resultCanBeNegative() {
        Expression expr = new SubtractExpression(new NumberExpression(2), new NumberExpression(5));
        assertEquals(-3.0, expr.interpret(new Context()), 0.001);
    }


    // ── MultiplyExpression ────────────────────────────────────
    @Test
    public void multiply_multipliesTwoNumbers() {
        Expression expr = new MultiplyExpression(new NumberExpression(4), new NumberExpression(5));
        assertEquals(20.0, expr.interpret(new Context()), 0.001);
    }

    @Test
    public void multiply_multipliesVariableByNumber() {
        Context ctx = new Context().set("a", 6);
        Expression expr = new MultiplyExpression(new VariableExpression("a"), new NumberExpression(7));
        assertEquals(42.0, expr.interpret(ctx), 0.001);
    }


    // ── DivideExpression ──────────────────────────────────────
    @Test
    public void divide_dividesLeftByRight() {
        Expression expr = new DivideExpression(new NumberExpression(10), new NumberExpression(2));
        assertEquals(5.0, expr.interpret(new Context()), 0.001);
    }

    @Test(expected = ArithmeticException.class)
    public void divide_throwsOnZero() {
        new DivideExpression(new NumberExpression(5), new NumberExpression(0)).interpret(new Context());
    }

    @Test
    public void divide_producesFractionalResult() {
        Expression expr = new DivideExpression(new NumberExpression(10), new NumberExpression(4));
        assertEquals(2.5, expr.interpret(new Context()), 0.001);
    }


    // ── Composite trees ───────────────────────────────────────
    @Test
    public void composite_diffOfSquares() {
        Expression a = new VariableExpression("a");
        Expression b = new VariableExpression("b");
        Context ctx  = new Context().set("a", 10).set("b", 5);
        Expression expr = new MultiplyExpression(new AddExpression(a, b), new SubtractExpression(a, b));
        assertEquals(75.0, expr.interpret(ctx), 0.001);
    }

    @Test
    public void composite_sameTreeDifferentContexts() {
        Expression a    = new VariableExpression("a");
        Expression b    = new VariableExpression("b");
        Expression expr = new MultiplyExpression(new AddExpression(a, b), new SubtractExpression(a, b));
        assertEquals(75.0, expr.interpret(new Context().set("a", 10).set("b", 5)), 0.001);
        assertEquals(32.0, expr.interpret(new Context().set("a",  6).set("b", 2)), 0.001);
    }

    @Test
    public void composite_literalThreeLevelTree() {
        Expression expr = new MultiplyExpression(
            new AddExpression(new NumberExpression(3), new NumberExpression(4)),
            new NumberExpression(2));
        assertEquals(14.0, expr.interpret(new Context()), 0.001);
    }
}
